bpf/bpf_sock.c

```diff
@@ -775,6 +780,11 @@ static __always_inline int __sock6_xlate_fwd(struct bpf_sock_addr *ctx,
 	bool backend_from_affinity = false;
 	__u32 backend_id = 0;
 
+#ifdef SKIP_SK_LB_IN_POD_NS
+	if (!in_hostns)
```

Comment:
Please keep in mind that availability of the `in_hostns` depends on the `get_netns_cookie()` kernel helper introduced in https://lore.kernel.org/bpf/cover.1585323121.git.daniel@iogearbox.net/T/. So, you might want to check for the helpers availability in the agent, and notify the user about its absence in the case they have enabled SkipSkLBInPodNS.

Reply:
I believe this is checked in `ctx_in_hostns()`. In the absence of such helpers, `in_hostns` will be set to `true` and we won't skip socket lb.

Reply:
Is the workaround ineffective on those particular kernels then? Or does this mean that host services will break when this flag is configured on those kernels?

If either of the above cases hold, it seems like that would be useful to notify the user about when they enable this flag and do not have the corresponding bpf helper.

Reply:
> Is the workaround ineffective on those particular kernels then

This. Therefore, I'm suggesting to warn users via log msg about Istio rules being bypasses. If we make the flag less generic, i.e. Istio specific, then instead of warning we could panic.

Reply:
Added logging in `daemon/cmd/kube_proxy_replacement.go`.

Reply:
I think this will result in odd semantics, meaning, for the BPF kube-proxy replacement we'll end up skipping everything for Pods, but not for host ns. The latter will be fine, but for Pods given we compile in xlation in bpf_lxc, we'll have ClusterIP services, and NodePort we'll pass through. The BPF code would then translate this on remote node via N-S LB, but for example externalIPs or hostPort translation for local node will not happen. This means this feature will really only work when you run the regular kube-proxy on the side to cover the remaining corner cases, is that intentional?

Reply:
Could you also update the kube-proxy-free guide with the concrete semantics? Probably good fit under: https://docs.cilium.io/en/v1.9/gettingstarted/kubeproxy-free/#kube-proxy-hybrid-modes

Reply:
hmm, I'm reviewing what we did/needs to be done for this PR now and realized I completely overlooked this comment somehow a while ago...so basically are you saying all "host-reachable" services for pod ns won't work if this option is turned on? @borkmann